sos/plugins/networking.py

```diff
@@ -155,6 +155,54 @@ def setup(self):
             pred=SoSPredicate(self, kmods=['ip6table_filter'])
         )
 
+        # Get devlink output
+        self.add_cmd_output([
+                    "devlink dev",
+                    "devlink dev param show",
+                    "devlink dev info",
+                    "devlink port",
+                    "devlink sb show",
+                    "devlink sb pool show",
+                    "devlink sb port pool show",
+                    "devlink sb tc bind show",
+                    "devlink -s -v trap show"
+                ])
+
+        # Collect information about devlink devices and dpipe tables
+        devlink_dev = self.exec_cmd("devlink dev")
+        if devlink_dev['status'] == 0:
+            devices = []
+            dev_tables = []
+            for line_dev in devlink_dev['output'].splitlines():
+                devices.append(line_dev)
+                # Collect information about devlink dpipe tables
+                devlink_dev_tables = self.exec_cmd(
+                    "devlink dpipe table show %s" % line_dev
+                    )
+                if devlink_dev_tables['status'] != 0:
+                    continue
+                for line_tables in devlink_dev_tables['output'].splitlines():
+                    line_tables = line_tables.strip()
+                    if line_tables.startswith('name'):
+                        dev_tables.append((line_dev,
+                                            line_tables.split()[1]))
+            # Capture per-device output from devlink
+            self.add_cmd_output(["devlink sb occupancy snapshot %s" % dev
```

Comment:
If you're having to split a call like this the preferred format is:

```python
    self.add_cmd_output([
        "devlink sb occupancy snapshot %s" % dev for dev in devices
    ])
```

That said, I think we can improve formatting and readability further here by using a table of commands:

```python
    devlink_cmds = [
        "devlink sb occupancy snapshot %s",
        "devlink sb occupancy show %s",
        ...
    ]

    self.add_cmd_output([cmd % dev for cmd in devlink_cmds for dev in devices])
```

E.g.:

```python
>>> cmds = ["foo %s", "bar %s", "baz %s", "qux %s"]
>>> devs = ["xuq", "zab", "rab", "oof"]
>>> [(cmd % dev) for cmd in cmds for dev in devs]
['foo xuq', 'foo zab', 'foo rab', 'foo oof', 'bar xuq', 'bar zab', 'bar rab', 'bar oof', 'baz xuq', 'baz zab', 'baz rab', 'baz oof', 'qux xuq', 'qux zab', 'qux rab', 'qux oof']
```

Reply:
Thank you! I rewrote this part.